sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -71,14 +91,49 @@ private[spark] object CosmosClientCache {
         }
 
         val client = builder.buildAsyncClient()
+        val epochNowInMs = Instant.now.toEpochMilli
+        val newClientCacheEntry = CosmosClientCacheMetadata(
+          client,
+          new AtomicLong(epochNowInMs),
+          new AtomicLong(epochNowInMs))
 
-        cache.putIfAbsent(cosmosClientConfiguration, client) match {
+        cache.putIfAbsent(cosmosClientConfiguration, newClientCacheEntry) match {
           case None =>
             client
           case Some(_) =>
             throw new ConcurrentModificationException("Should not reach here because its synchronized")
         }
     }
+  }
+  // scalastyle:on method.length
 
+  private[this] def onCleanup(): Unit = {
+    try {
+      logInfo(s"-->onCleanup (${cache.size} clients)")
+      val snapshot = cache.foreach((pair) => {
+        val clientConfig = pair._1
+        val clientMetadata = pair._2
+
+        if (clientMetadata.lastRetrieved.get() < Instant.now.toEpochMilli - unusedClientTtlInMs) {
+          // Only remove the client from the cache - don't purge or close!!! (because the client retrieved 15 minutes
+          // ago might still be used
+          logInfo(s"Removing client due to inactivity from the cache - ${clientConfig.endpoint}, " +
+            s"${clientConfig.applicationName}, ${clientConfig.preferredRegionsList}, ${clientConfig.useGatewayMode}, " +
+            s"${clientConfig.useEventualConsistency}")
+          cache.remove(clientConfig)
```

Comment:
if we only remove it from the cache, how it will help the memory issue?

Reply:
GC will be able to clean it up because it isn't referenced/rooted anymore. I have validated that the number of Instances of CosmosAsyncClient in the heap dumps reduced and settles on a stable number